pkg/virtctl/create/vm.go

```diff
@@ -0,0 +1,242 @@
+ * This file is part of the KubeVirt project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * Copyright 2022 Red Hat, Inc.
+ *
+ */
+
+package create
+
+import (
+	"github.com/spf13/cobra"
+	k8sv1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/rand"
+	v1 "kubevirt.io/api/core/v1"
+	"kubevirt.io/api/instancetype"
+	"sigs.k8s.io/yaml"
+
+	"kubevirt.io/kubevirt/pkg/virtctl/templates"
+)
+
+const (
+	VM = "vm"
+
+	NameFlag                   = "name"
+	RunStrategyFlag            = "run-strategy"
+	InstancetypeFlag           = "instancetype"
+	ClusterInstancetypeFlag    = "cluster-instancetype"
+	PreferenceFlag             = "preference"
+	ClusterPreferenceFlag      = "cluster-preference"
+	PvcFlag                    = "pvc"
+	DatavolumeFlag             = "datavolume"
+	TerminationGracePeriodFlag = "termination-grace-period"
+)
+
+type createVM struct {
+	name                   string
+	instancetype           string
+	preference             string
+	pvc                    string
+	dataVolume             string
+	runStrategy            string
+	terminationGracePeriod int64
+}
+
+type optionFn func(*createVM, *v1.VirtualMachine)
+
+var optFns = map[string]optionFn{
+	InstancetypeFlag:        withInstancetype,
+	ClusterInstancetypeFlag: withClusterInstancetype,
+	PreferenceFlag:          withPreference,
+	ClusterPreferenceFlag:   withClusterPreference,
+	PvcFlag:                 withPVC,
+	DatavolumeFlag:          withDataVolume,
+}
+
+func NewVirtualMachineCommand() *cobra.Command {
+	c := defaultCreateVM()
+	cmd := &cobra.Command{
+		Use:     VM,
+		Short:   "Create a VirtualMachine manifest.",
+		Example: c.usage(),
+		RunE: func(cmd *cobra.Command, _ []string) error {
+			return c.run(cmd)
+		},
+	}
+
+	cmd.Flags().StringVar(&c.name, NameFlag, c.name, "Specify the name of the VirtualMachine.")
+
+	cmd.Flags().StringVar(&c.instancetype, InstancetypeFlag, c.instancetype, "Specify the Instancetype of the VirtualMachine.")
+	cmd.Flags().StringVar(&c.instancetype, ClusterInstancetypeFlag, c.instancetype, "Specify the ClusterInstancetype of the VirtualMachine.")
+	cmd.MarkFlagsMutuallyExclusive(InstancetypeFlag, ClusterInstancetypeFlag)
+
+	cmd.Flags().StringVar(&c.preference, PreferenceFlag, c.preference, "Specify the Preference of the VirtualMachine.")
+	cmd.Flags().StringVar(&c.preference, ClusterPreferenceFlag, c.preference, "Specify the ClusterPreference of the VirtualMachine.")
```

Comment:
rather than two options, i wonder if it would make sense to have single preference option where the kind can be included to specify if it's a cluster or namespaced scoped preference?

something like
`--preference=[optional kind/]<preference>`

so, for a namespaced preference
`--preference=VirtualMachinePreference/mypref` 

for a global preference
`--preference=VirtualMachineClusterPreference/mypref` 

and default to global scoped preference when no kind delimiter is used
`--preference=mypref`

Reply:
Addressed this.